gnovm/pkg/doc/dirs.go

```diff
@@ -0,0 +1,163 @@
+// Mostly copied from go source at tip, commit d922c0a.
+//
+// Copyright 2015 The Go Authors. All rights reserved.
+
+package doc
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// A Dir describes a directory holding code by specifying
+// the expected import path and the file system directory.
+type Dir struct {
+	importPath string // import path for that dir
+	dir        string // file system directory
+}
+
+// Dirs is a structure for scanning the directory tree.
+// Its Next method returns the next Go source directory it finds.
+// Although it can be used to scan the tree multiple times, it
+// only walks the tree once, caching the data it finds.
+type Dirs struct {
+	scan   chan Dir // Directories generated by walk.
+	hist   []Dir    // History of reported Dirs.
+	offset int      // Counter for Next.
+}
+
+// NewDirs begins scanning the given stdlibs directory.
+func NewDirs(dirs ...string) *Dirs {
+	d := &Dirs{
+		hist: make([]Dir, 0, 256),
```

Comment:
Eh, it's a sort of "average-case guess" for how many directories there will be. I mostly add these because starting off without any just makes [Go realloc the slice every power of two when appending](https://go.dev/play/p/Wut3p3eMeRm) -- so it's more of a way of hinting "it'll probably grow at least to a few hundred".